src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs

```diff
@@ -0,0 +1,218 @@
+    }
+
+    private async Task UpdateInstanceFileAsync(InstanceFileIdentifier instanceFileIdentifier, DicomDataset datasetToUpdate, CancellationToken cancellationToken)
+    {
+        long originFileIdentifier = instanceFileIdentifier.Version;
+        long newFileIdentifier = instanceFileIdentifier.NewVersion.Value;
+        var stopwatch = new Stopwatch();
+
+        KeyValuePair<string, long> block = default;
+
+        // If the file is already updated, then we can copy the file and just update the first block
+        if (instanceFileIdentifier.OriginalVersion.HasValue)
+        {
+            _logger.LogInformation("Begin copying instance file {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+            await _fileStore.CopyFileAsync(originFileIdentifier, newFileIdentifier, cancellationToken);
+        }
+        else
+        {
+            stopwatch.Start();
+            _logger.LogInformation("Begin downloading original file {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+
+            using Stream stream = await _fileStore.GetFileAsync(originFileIdentifier, cancellationToken);
+            DicomFile dcmFile = await DicomFile.OpenAsync(stream, FileReadOption.ReadLargeOnDemand, LargeObjectSizeInBytes);
+
+            long firstBlockLength = stream.Length;
+
+            // If the file is greater than 1MB try to upload in different blocks.
+            // Since we are supporting updating only Patient demographic and identification module,
+            // we assume patient attributes are at the very beginning of the dataset.
+            // If in future, we support updating other attributes like private tags, which could appear at the end
+            // We need to read the whole file instead of first block.
+            if (stream.Length > _largeDicomItemSizeInBytes)
+            {
+                _logger.LogInformation("Found bigger DICOM item, splitting the file into multiple blocks. {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+
+                if (dcmFile.Dataset.TryGetLargeDicomItem(_largeDicomItemSizeInBytes, _stageBlockSizeInBytes, out DicomItem largeDicomItem))
+                {
+                    RemoveItemsAfter(dcmFile.Dataset, largeDicomItem.Tag);
+                }
+
+                firstBlockLength = await DicomFileExtensions.GetByteLengthAsync(dcmFile, _recyclableMemoryStreamManager);
+            }
+
+            IDictionary<string, long> blockLengths = GetBlockLengths(stream.Length, firstBlockLength);
+
+            _logger.LogInformation("Begin uploading instance file in blocks {OrignalFileIdentifier} - {NewFileIdentifier}", originFileIdentifier, newFileIdentifier);
+
+            stream.Seek(0, SeekOrigin.Begin);
+
+            await _fileStore.StoreFileInBlocksAsync(newFileIdentifier, stream, blockLengths, cancellationToken);
+
+            block = blockLengths.First();
```

Comment:
Isn't this always non-`null` because `blockLengths` always have count >= 1

Reply:
Yes, this is always not null when we are updating first time, but when we already updated, we wont go through this code. So at that time, it will be null

Reply:
Oh my gosh, I didn't notice the bracket!